File: app/src/main/java/eu/pkgsoftware/babybuddywidgets/BabyLayoutHolder.java
package eu.pkgsoftware.babybuddywidgets;

import android.view.View;

import com.squareup.phrase.Phrase;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import eu.pkgsoftware.babybuddywidgets.activitycomponents.TimerControl;
import eu.pkgsoftware.babybuddywidgets.timers.FragmentCallbacks;
import eu.pkgsoftware.babybuddywidgets.timers.LoggingButtonController;
import eu.pkgsoftware.babybuddywidgets.databinding.BabyManagerBinding;
import eu.pkgsoftware.babybuddywidgets.history.ChildEventHistoryLoader;
import eu.pkgsoftware.babybuddywidgets.networking.BabyBuddyClient;
import eu.pkgsoftware.babybuddywidgets.networking.ChildrenStateTracker;
import eu.pkgsoftware.babybuddywidgets.networking.babybuddy.models.TimeEntry;
import eu.pkgsoftware.babybuddywidgets.timers.TimerControlInterface;
import eu.pkgsoftware.babybuddywidgets.timers.TimersUpdatedCallback;
import eu.pkgsoftware.babybuddywidgets.timers.TranslatedException;
import eu.pkgsoftware.babybuddywidgets.utils.Promise;

public class BabyLayoutHolder extends RecyclerView.ViewHolder implements TimerControlInterface {
    private final BabyManagerBinding binding;
    private final BaseFragment baseFragment;
    private final BabyBuddyClient client;


    private BabyBuddyClient.Child child = null;

    private ChildEventHistoryLoader childHistoryLoader = null;

    private ChildrenStateTracker.ChildObserver childObserver = null;

    private BabyBuddyClient.Timer[] cachedTimers = null;
    private List<TimersUpdatedCallback> updateTimersCallbacks = new ArrayList<>(10);

    private int pendingTimerModificationCalls = 0;

    private LoggingButtonController loggingButtonController = null;

    public BabyLayoutHolder(BaseFragment fragment, BabyManagerBinding bmb) {
        super(bmb.getRoot());
        binding = bmb;

        baseFragment = fragment;
        client = fragment.getMainActivity().getClient();

        binding.mainScrollView.setOnScrollChangeListener((v, scrollX, scrollY, oldScrollX, oldScrollY) -> {
            if (childHistoryLoader != null) {
                childHistoryLoader.updateTop();
            }
        });
    }

    public BabyBuddyClient.Child getChild() {
        return child;
    }

    private void requeueImmediateTimerListRefresh() {
        client.listTimers(child.id, new BabyBuddyClient.RequestCallback<BabyBuddyClient.Timer[]>() {
            @Override
            public void error(@NonNull Exception error) {
            }

            @Override
            public void response(BabyBuddyClient.Timer[] response) {
                updateTimerList(response);
            }
        });
    }

    private void resetChildHistoryLoader() {
        if (childHistoryLoader != null) {
            childHistoryLoader.close();
        }
        childHistoryLoader = null;
    }

    public void updateChild(BabyBuddyClient.Child c, ChildrenStateTracker stateTracker) {
        clear();
        this.child = c;

        if (child != null) {
            if (stateTracker == null) {
                throw new RuntimeException("StateTracker was null somehow");
            }
            childObserver = stateTracker.new ChildObserver(child.id, this::updateTimerList);

            childHistoryLoader = new ChildEventHistoryLoader(
                baseFragment,
                binding.innerTimeline,
                child.id,
                new VisibilityCheck(binding.mainScrollView),
                binding.timelineProgressSpinner,
                (entryType, exception) -> {
                    String tActivity = baseFragment.translateActivityName(entryType);

                    String msg = Phrase.from(baseFragment.getResources(), R.string.history_loading_timeline_entry_failed)
                        .put("activity", tActivity)
                        .format().toString();

                    baseFragment.getMainActivity().binding.globalErrorBubble.flashMessage(msg);
                }

            );

            loggingButtonController = new LoggingButtonController(
                baseFragment,
                binding,
                new FragmentCallbacks() {
                    @Override
                    public void insertControls(@NonNull View view) {
                        if (view.getParent() != null) {
                            return;
                        }
                        binding.loggingEditors.addView(view);
                    }

                    @Override
                    public void removeControls(@NonNull View view) {
                        if (view.getParent() == null) {
                            return;
                        }
                        binding.loggingEditors.removeView(view);
                    }

                    @Override
                    public void updateTimeline(@Nullable TimeEntry newEntry) {
                        if (childHistoryLoader != null) {
                            if (newEntry != null) {
                                childHistoryLoader.addEntryToTop(newEntry);
                            }
                            childHistoryLoader.forceRefresh();
                        }
                    }
                },
                child,
                this
            );
        }
    }

    public void updateTimerList(BabyBuddyClient.Timer[] timers) {
        if (pendingTimerModificationCalls > 0) {
            // Buffer timer-list updates while a timer-modifying operation is running
            // (prevents confusing UI updates)
            return;
        }

        if (child == null) {
            cachedTimers = new BabyBuddyClient.Timer[0];
            callTimerUpdateCallback();
            return;
        }

        for (BabyBuddyClient.Timer t : timers) {
            if (t.child_id != child.id) {
                return;
            }
        }

        cachedTimers = timers;
        callTimerUpdateCallback();
    }

    public void onViewDeselected() {
        if (loggingButtonController != null) {
            loggingButtonController.storeStateForSuspend();
        }
        resetChildObserver();
        resetChildHistoryLoader();
    }

    private void resetChildObserver() {
        if (childObserver != null) {
            childObserver.close();
            childObserver = null;
        }
    }

    public void clear() {
        if (loggingButtonController != null) {
            loggingButtonController.storeStateForSuspend();
            loggingButtonController.destroy();
            loggingButtonController = null;
        }
        resetChildObserver();
        resetChildHistoryLoader();
        child = null;
        cachedTimers = null;
    }

    public void close() {
        clear();
    }

    private class UpdateBufferingPromise<A, B> implements Promise<A, B> {
        private Promise<A, B> promise;

        public UpdateBufferingPromise(Promise<A, B> promise) {
            this.promise = promise;
            pendingTimerModificationCalls++;
        }

        @Override
        public void succeeded(A a) {
            pendingTimerModificationCalls--;
            promise.succeeded(a);
        }

        @Override
        public void failed(B b) {
            pendingTimerModificationCalls--;
            promise.failed(b);
        }
    }

    @Override
    public void createNewTimer(@NonNull BabyBuddyClient.Timer timer, @NonNull Promise<BabyBuddyClient.Timer, TranslatedException> cb) {
        baseFragment.getMainActivity().getChildTimerControl(child).createNewTimer(
            timer, new UpdateBufferingPromise<>(cb)
        );
    }

    @Override
    public void startTimer(@NotNull BabyBuddyClient.Timer timer, @NonNull Promise<BabyBuddyClient.Timer, TranslatedException> cb) {
        baseFragment.getMainActivity().getChildTimerControl(child).startTimer(
            timer, new UpdateBufferingPromise<>(cb)
        );
    }

    @Override
    public void stopTimer(@NotNull BabyBuddyClient.Timer timer, @NonNull Promise<Object, TranslatedException> cb) {
        baseFragment.getMainActivity().getChildTimerControl(child).stopTimer(
            timer, new UpdateBufferingPromise<>(cb)
        );
    }

    @Override
    public void registerTimersUpdatedCallback(@NonNull TimersUpdatedCallback callback) {
        if (updateTimersCallbacks.contains(callback)) {
            return;
        }
        updateTimersCallbacks.add(callback);

        baseFragment.getMainActivity().getChildTimerControl(child).registerTimersUpdatedCallback(
            timers -> {
                for (TimersUpdatedCallback c : updateTimersCallbacks) {
                    c.newTimerListLoaded(timers);
                }
            }
        );
        callTimerUpdateCallback();
    }

    @Override
    public void unregisterTimersUpdatedCallback(@NonNull TimersUpdatedCallback callback) {
        if (!updateTimersCallbacks.contains(callback)) {
            return;
        }
        updateTimersCallbacks.remove(callback);
    }

    private void callTimerUpdateCallback() {
        // urgh...
        TimerControl wrapped = (TimerControl) baseFragment.getMainActivity().getChildTimerControl(child).getWrap();
        if (cachedTimers != null) {
            wrapped.callTimerUpdateCallback(cachedTimers);
        }
    }

    @NonNull
    @Override
    public CredStore.Notes getNotes(@NonNull BabyBuddyClient.Timer timer) {
        return baseFragment.getMainActivity().getChildTimerControl(child).getNotes(timer);
    }

    @Override
    public void setNotes(@NonNull BabyBuddyClient.Timer timer, CredStore.Notes notes) {
        baseFragment.getMainActivity().getChildTimerControl(child).setNotes(timer, notes);
    }
}

File: app/src/main/java/eu/pkgsoftware/babybuddywidgets/BabyLayoutHolder.kt
package eu.pkgsoftware.babybuddywidgets

import android.view.View
import androidx.recyclerview.widget.RecyclerView
import com.squareup.phrase.Phrase
import eu.pkgsoftware.babybuddywidgets.activitycomponents.TimerControl
import eu.pkgsoftware.babybuddywidgets.databinding.BabyManagerBinding
import eu.pkgsoftware.babybuddywidgets.history.ChildEventHistoryLoader
import eu.pkgsoftware.babybuddywidgets.history.ShowErrorPill
import eu.pkgsoftware.babybuddywidgets.networking.BabyBuddyClient
import eu.pkgsoftware.babybuddywidgets.networking.babybuddy.models.Child
import eu.pkgsoftware.babybuddywidgets.networking.babybuddy.models.TimeEntry
import eu.pkgsoftware.babybuddywidgets.timers.FragmentCallbacks
import eu.pkgsoftware.babybuddywidgets.timers.LoggingButtonController
import eu.pkgsoftware.babybuddywidgets.timers.TimerControlInterface
import eu.pkgsoftware.babybuddywidgets.timers.TimersUpdatedCallback
import eu.pkgsoftware.babybuddywidgets.timers.TranslatedException
import eu.pkgsoftware.babybuddywidgets.utils.Promise

class BabyLayoutHolder(
    private val baseFragment: BaseFragment,
    private val binding: BabyManagerBinding
) : RecyclerView.ViewHolder(
    binding.getRoot()
), TimerControlInterface {
    private val client: BabyBuddyClient

    var child: Child? = null

    private var childHistoryLoader: ChildEventHistoryLoader? = null

    private var cachedTimers: Array<BabyBuddyClient.Timer>? = null
    private val updateTimersCallbacks: MutableList<TimersUpdatedCallback> =
        ArrayList(10)

    private var pendingTimerModificationCalls = 0

    private var loggingButtonController: LoggingButtonController? = null

    init {
        client = baseFragment.mainActivity.client

        binding.mainScrollView.setOnScrollChangeListener(View.OnScrollChangeListener { v: View?, scrollX: Int, scrollY: Int, oldScrollX: Int, oldScrollY: Int ->
            if (childHistoryLoader != null) {
                childHistoryLoader!!.updateTop()
            }
        })
    }

    private fun resetChildHistoryLoader() {
        if (childHistoryLoader != null) {
            childHistoryLoader!!.close()
        }
        childHistoryLoader = null
    }

    fun updateChild(c: Child) {
        if (child === c) {
            return
        }

        clear()
        this.child = c

        child ?.let { child ->
            childHistoryLoader = ChildEventHistoryLoader(
                baseFragment,
                binding.innerTimeline,
                child.id,
                VisibilityCheck(binding.mainScrollView),
                binding.timelineProgressSpinner,
                object : ShowErrorPill {
                    override fun showErrorPill(entryType: String, exception: Exception?) {
                        val tActivity = baseFragment.translateActivityName(entryType!!)
                        val msg = Phrase.from(
                            baseFragment.getResources(),
                            R.string.history_loading_timeline_entry_failed
                        )
                            .put("activity", tActivity)
                            .format().toString()
                        baseFragment.mainActivity.binding.globalErrorBubble.flashMessage(msg)
                    }
                }
            )

            loggingButtonController = LoggingButtonController(
                baseFragment,
                binding,
                object : FragmentCallbacks {
                    override fun insertControls(view: View) {
                        if (view.getParent() != null) {
                            return
                        }
                        binding.loggingEditors.addView(view)
                    }

                    override fun removeControls(view: View) {
                        if (view.getParent() == null) {
                            return
                        }
                        binding.loggingEditors.removeView(view)
                    }

                    override fun updateTimeline(newEntry: TimeEntry?) {
                        if (childHistoryLoader != null) {
                            if (newEntry != null) {
                                childHistoryLoader!!.addEntryToTop(newEntry)
                            }
                            childHistoryLoader!!.forceRefresh()
                        }
                    }
                },
                child,
                this
            )
        }
    }

    fun updateTimerList(timers: Array<BabyBuddyClient.Timer>) {
        if (pendingTimerModificationCalls > 0) {
            // Buffer timer-list updates while a timer-modifying operation is running
            // (prevents confusing UI updates)
            return
        }

        if (child == null) {
            cachedTimers = arrayOf()
            callTimerUpdateCallback()
            return
        }

        for (t in timers) {
            child?.let { child ->
                if (t.child_id != child.id) {
                    return
                }
            } ?: return
        }

        cachedTimers = timers
        callTimerUpdateCallback()
    }

    fun onViewDeselected() {
        if (loggingButtonController != null) {
            loggingButtonController!!.storeStateForSuspend()
        }
        resetChildHistoryLoader()
    }

    fun clear() {
        if (loggingButtonController != null) {
            loggingButtonController!!.storeStateForSuspend()
            loggingButtonController!!.destroy()
            loggingButtonController = null
        }
        resetChildHistoryLoader()
        child = null
        cachedTimers = null
    }

    fun close() {
        clear()
    }

    private inner class UpdateBufferingPromise<A, B>(private val promise: Promise<A, B>) :
        Promise<A, B> {
        init {
            pendingTimerModificationCalls++
        }

        override fun succeeded(a: A) {
            pendingTimerModificationCalls--
            promise.succeeded(a)
        }

        override fun failed(b: B) {
            pendingTimerModificationCalls--
            promise.failed(b)
        }
    }

    override fun createNewTimer(
        timer: BabyBuddyClient.Timer,
        cb: Promise<BabyBuddyClient.Timer, TranslatedException>
    ) {
        child?.let { child ->
            baseFragment.mainActivity.getChildTimerControl(child.id).createNewTimer(
                timer, UpdateBufferingPromise(cb)
            )
        }
    }

    override fun startTimer(
        timer: BabyBuddyClient.Timer,
        cb: Promise<BabyBuddyClient.Timer, TranslatedException>
    ) {
        child?.let { child ->
            baseFragment.mainActivity.getChildTimerControl(child.id).startTimer(
                timer, UpdateBufferingPromise(cb)
            )
        }
    }

    override fun stopTimer(timer: BabyBuddyClient.Timer, cb: Promise<Any, TranslatedException>) {
        child?.let { child ->
            baseFragment.mainActivity.getChildTimerControl(child.id).stopTimer(
                timer, UpdateBufferingPromise(cb)
            )
        }
    }

    override fun registerTimersUpdatedCallback(callback: TimersUpdatedCallback) {
        child?.let { child ->
            if (updateTimersCallbacks.contains(callback)) {
                return
            }
            updateTimersCallbacks.add(callback)

            baseFragment.mainActivity.getChildTimerControl(child.id).registerTimersUpdatedCallback(
                object : TimersUpdatedCallback {
                    override fun newTimerListLoaded(timers: Array<BabyBuddyClient.Timer>) {
                        for (c in updateTimersCallbacks) {
                            c.newTimerListLoaded(timers)
                        }
                    }
                }
            )

            callTimerUpdateCallback()
        }
    }

    override fun unregisterTimersUpdatedCallback(callback: TimersUpdatedCallback) {
        if (!updateTimersCallbacks.contains(callback)) {
            return
        }
        updateTimersCallbacks.remove(callback)
    }

    private fun callTimerUpdateCallback() {
        // urgh...
        child?.let { child ->
            val wrapped = baseFragment.mainActivity.getChildTimerControl(child.id).wrap as TimerControl
            if (cachedTimers != null) {
                wrapped.callTimerUpdateCallback(cachedTimers!!)
            }
        }
    }

    override fun getNotes(timer: BabyBuddyClient.Timer): CredStore.Notes {
        return child?.let { child ->
            return@let baseFragment.mainActivity.getChildTimerControl(child.id).getNotes(timer)
        } ?: CredStore.Notes("", false)
    }

    override fun setNotes(timer: BabyBuddyClient.Timer, notes: CredStore.Notes?) {
        child?.let { child ->
            baseFragment.mainActivity.getChildTimerControl(child.id).setNotes(timer, notes)
        }
    }
}
